012-OFEP-variations-and-variables.md

```diff
@@ -0,0 +1,167 @@
+
+This is where we introduce the concept of **Variations** and **Variables**.
+
+## Kinds of evaluations
+
+### Flags
+
+This is solely for checking if the feature flag itself is enabled or not, and is always of `Boolean` type.
+
+Example:
+
+```js
+const featureKey = "my_feature";
+const defaultState = false;
+
+const isEnabled = client.isEnabled(featureKey, defaultState);
+```
+
+### Variations
+
+We now enter the realm of A/B testing and experimentation.
+
+An experiment consists of different variations (which are usually of `String` type), and the user is consistently bucketed (assigned) into one of its variations, irrespective of the device or session the user is in.
+
+You can think of an experiment named `my_experiment`, that can have two variations:
+
+- `control`: the default behaviour
+- `treatment`: the new behaviour that we wish to test
+
+A user will be bucketed into one of these variations, and will stay in that variation for the duration of the experiment.
+
+Example:
+
+```js
+const featureKey = "my_experiment";
+const defaultVariation = "control";
+
+const variation = client.getVariation(featureKey, defaultVariation);
+```
+
+#### Weights
+
+Each variation is also associated with a weight, which is a number between 0 (0%) and 1 (100%), and represents the probability of the user being bucketed into that variation. The total weight of all variations in an experiment must be 1.
+
+This level of details is something to be handled at Provider level, and not the OpenFeature SDKs.
+
+#### Measuring success
+
+Applications will be tracking various different conversion goals (like successful user sign ups, purchases, etc). Based on the variation the user is bucketed into, we will be able to track the conversion goal events against each variation of the experiment over a period of time, and decide which variation is the winner.
+
+OpenFeature SDKs will not be involved here.
+
+### Variables
+
+Feature management can go beyond just on/off values for the flag itself and supporting a mechanism for a/b testing with variations.
+
+We may also want to support the concept of variables for supporting configuration to be deployed independently of our applications, which are values that can be changed over time.
+
+These values can be of various different primitive types for different use cases.
+
+You can think of a feature named `my_feature`, that can have several variables scoped under it:
+
+- `showSidebar`: boolean
+- `buttonColor`: string
+- `allowSignup`: boolean
+
+The variable values can be evaluated against the context that already exists in the SDK, and can also react further based on the bucketed variation if there's an experiment involved.
+
+Example:
+
+```js
+const featureKey = "my_feature";
+const variableKey = "showSidebar";
+const defaultValue = false;
+
+const showSidebar = client.getVariableBoolean(featureKey, variableKey, defaultValue);
```

Comment:
I like the concept, especially with the type enforcement at the variable level. We can do something similar, but your proposal may be a nicer dev experience.

Here's what's already possible.

```ts
const featureKey = "my_feature";
const defaultValue = { showSidebar: false };

const { showSidebar } = client.getObjectValue(featureKey, defaultValue);
```

In TypeScript, it would even be possible to register a hook that validates the object before returning it to the client.

```ts
import { z } from "zod";

const myFeatureSchema= z.object({
  showSidebar: z.boolean(),
});
type MyFeature = z.infer<typeof myFeatureSchema>;

const featureKey = "my_feature";
const defaultValue: MyFeature = { showSidebar: false };

class Validator {
	constructor(private readonly zodSchema) {}

	after(_, evaluationDetails): {
		this.schema.parse(evaluationDetails.value);
	}
}

const { showSidebar } = client.getObjectValue<MyFeature >(featureKey, defaultValue, undefined, { hooks: [new Validator(myFeatureSchema)] });
```
> NOTE: This was written from memory and has yet to be tested.

The second example is fairly complex but would offer runtime variable validation.

Reply:
`client.getObject()` can work, but this will require the provider to compute all the variables together and return it as an object, which application can destruct from.

the proposed API in document suggests we evaluate individual variables scoped under each feature on-demand.

also when the method name specifically says `getVariable*`, it shows the intent and the kind of desired value as well. We want the "variable" to be evaluated with that specific key.

Reply:
I think I had discussed this in another OFEP or spec comment, but I think that the idea of a "scoped" feature could be added and would probably be more widely applicable. 

Variables for this system seem closer to what variations may generally be from other systems (multiple types, potentially objects), and you would use several "flags" to accomplish something similar in those systems (or one structured flag containing the relevant data).

Here they are together in that a flag contains a few more states that are exposed.

If you were to imagine this at a very basic level it would be something like.

"myFeature" -> Variation (with on/off as details of that evaluation). Maybe this means some extra work is done.

"myFeature:showSideBar" -> Variable.

Being strings is for illustrative purpose, but if there is true need, then some more specific way could be devoted to support the idea.

Currently I think this is divergent enough that it would create a lot of dead API surface for most providers. There may be a way to make it into something more uniform though. How do we adjust the API so that two providers are not completely divergent from an API usage perspective?

Reply:
> If you were to imagine this at a very basic level it would be something like.
> 
> "myFeature" -> Variation (with on/off as details of that evaluation). Maybe this means some extra work is done.
> 
> "myFeature:showSideBar" -> Variable.
> 
> Being strings is for illustrative purpose, but if there is true need, then some more specific way could be devoted to support the idea.

This is an interesting idea.

> Currently I think this is divergent enough that it would create a lot of dead API surface for most providers.

I haven't had time yet to really dig into this proposal, but I think as it stands, this is my biggest concern with it.

Reply:
>another challenge would be emitting an "activation" event when an Experiment has been exposed to end users.

There is PROVIDER_CONFIGURATION_CHANGED [event](https://github.com/open-feature/spec/blob/main/specification/sections/05-events.md#requirement-511) , you can use it with [event details](https://github.com/open-feature/spec/blob/main/specification/types.md#provider-event-details) to reflect it. Can this solve your challenge ?

Reply:
The challenge is primarily on the triggering side, not listening side.

For listening, I am expecting Provider can also accept an option like this:

```js
OpenFeature.setProvider(new FeaturevisorProvider({
  onActivation(featureKey, variationValue) {
    // do tracking here
  }
}));
```

It's the triggering part, which happens via Featurevisor SDK instance like this:

```js
const instance; // Featurevisor SDK instance

// .activate() is like .getVariation(), 
// but it also bubbles an event up for listening
const variation = instance.activate("myFeatureKey", context);
```

SDK docs for activation: https://featurevisor.com/docs/sdks/#activation

Based on the comment thread here, the evaluation of a feature's variation could happen like this with OpenFeature:

```js
const variation = client.getStringValue("myFeatureKey:variation", "control", context);
```

Which is equivalent to Featurevisor SDK's `.getVariation()` method: https://featurevisor.com/docs/sdks/#getting-variations

But what could be an idiomatic way of calling the `.activate()` method with OpenFeature?

Reply:
Internally for event management, it can potentially solve the case.

but which API method of OpenFeature client to use in this case? what API call from the OpenFeature SDK client instance should trigger the "activate" event, knowing that we aren't extending the client API with more methods besides existing `get*Value()` methods?

usually in a/b testing libraries, the activation event is intentionally triggered by the application when it is known that the user has been exposed to a certain experiment (feature in our case) once. we do not trigger it right when we evaluate the value. it's a matter of visibility here.

that's where I am struggling to fit the OpenFeature model here.

Reply:
Maybe by "activating" you are referring to flag configuration change, like other changes of changing flag values for example from disabled to enabled ?
As far as I know, the OpenFeature client is dealing with application reading the flags values, not editing the flags configuration, as editing flags configurations and values is done by the feature flags service (e.g. FeatureVisor).
Is that what you mean, or can you elaborate ?

Reply:
I will try to summarize the differences below:

## 1: configuration update

this is what happens when the fetched configuration (Featurevisor calls them datafiles) has new changes, and the evaluated values can change further.

links:

- https://featurevisor.com/docs/building-datafiles/
- https://featurevisor.com/docs/sdks/#refreshing-datafile
- https://featurevisor.com/docs/sdks/#events
- https://featurevisor.com/docs/integrations/partykit/

## 2: feature activations

this happens when we wish to track when a particular user has been bucketed into a specific variation of an experiment. 

we intentionally activate the feature/experiment from the client, it bubbles an event up to our listener, and then we send the event to a tracker of our choice (could be Google Analytics, SegmentIO, or anything else)

links

- https://featurevisor.com/docs/use-cases/experiments/
- https://featurevisor.com/docs/sdks/#activation
- https://featurevisor.com/docs/bucketing/

Activation does not change the configuration in anyway.

Reply:
You could bubble up events using hooks. A hook can be configured to run after each flag evaluation and even be included in the provider itself. It's also possible to pass "hook hints" at evaluation time. These can be used in a hook to customize the behavior even further.

Reply:
> A hook can be configured to run after each flag evaluation

this is what I am specifically trying to avoid. we should not emit an event at the time of evaluation.

the event should be emitted only when the application thinks the user has been exposed to that feature/experiment, manually.

Reply:
> Hey just adding my voice to this as we @ DevCycle have a somewhat similar hierarchy to what's described here, having Features as our top-level organizational object that contain multiple variables (flags) underneath: https://docs.devcycle.com/introduction/core-concepts/feature-hierarchy https://docs.devcycle.com/sdk/server-side-sdks/node/node-openfeature https://github.com/DevCycleHQ/js-sdks/blob/main/sdk/openfeature-nodejs-provider/src/DevCycleProvider.ts
> 
> We generally have learned that the vast majority of folks want to just interface with flags (variables) as the main interface to our SDKs. There are very few folks who need the information about the feature + variation data at run-time, mostly that data is just needed for reporting / eventing. That is certainly something we've struggled with ourselves with OpenFeature is how to get the variation + feature metadata attached to the resolution details.
> 
> I think a better approach here would be to figure out how to keep the core interface of the OpenFeature spec the same, where users are interfacing with different types of flags (variables), but exposing the feature + variation data as part of the Resolution Details / [Detailed Evaluation](https://openfeature.dev/docs/reference/concepts/evaluation-api#detailed-evaluation) data.

This is an interesting perspective, and an appealing idea, to me at least.

@fahad19 do you think we could add well-defined properties to the [detailed evaluation](https://openfeature.dev/specification/types#evaluation-details) methods that would help? `value` would still maintain its current semantics (variable in DevCycle language as @jonathannorris mentions), but we could perhaps add additional properties that would satisfy your use-case, while remaining coherent with systems that don't have such concepts.

Reply:
thanks @toddbaert and @jonathannorris for all the inputs <3

I don't think the discussion here regarding activation will go anywhere, except for pushing it to the userland entirely.

I will close it with a summarized comment in a bit.